### PR TITLE
Ray Query support

### PR DESCRIPTION
**Pull request by kvark**
Path support:
- [x] WGSL-in
- [x] validation
- [x] SPV-out
- [x] MSL-out
- [ ] HLSL-out (not planned for this PR)

Note: PR also adds basic infra for specially generated types in IR, which may be handy for #1161

### CODE REVIEW

File: src/valid/interface.rs [hunk 1]
```diff
@@ -440,7 +440,9 @@ impl super::Validator {
                 match types[var.ty].inner {
                     crate::TypeInner::Image { .. }
                     | crate::TypeInner::Sampler { .. }
-                    | crate::TypeInner::BindingArray { .. } => {}
+                    | crate::TypeInner::BindingArray { .. }
+                    | crate::TypeInner::AccelerationStructure
+                    | crate::TypeInner::RayQuery => {}
```
Comment (jimblandy) [inline src/valid/interface.rs:445]: Can `RayQuery` handles really be global? I thought they were always local variables.

File: src/valid/expression.rs [hunk 2]
```diff
@@ -1427,6 +1427,7 @@ impl super::Validator {
                     return Err(ExpressionError::InvalidArrayType(expr));
                 }
             },
+            E::RayQueryProceedResult | E::RayQueryGetIntersection { .. } => ShaderStages::all(),
```
Comment (jimblandy) [inline src/valid/expression.rs:1430]: Can we fill in the type checks here?

File: src/front/wgsl/lower/mod.rs [hunk 3]
```diff
@@ -2378,4 +2434,28 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
...
+        let span = ctx.ast_expressions.get_span(expr);
+        let pointer = self.expression(expr, ctx.reborrow())?;
+
+        ctx.grow_types(pointer)?;
+        match *ctx.resolved_inner(pointer) {
+            crate::TypeInner::Pointer { base, .. } => match ctx.module.types[base].inner {
+                crate::TypeInner::RayQuery => Ok(pointer),
+                ref other => {
+                    log::error!("Pointer type to {:?} passed to ray query op", other);
+                    Err(Error::InvalidAtomicPointer(span))
```
Comment (jimblandy) [inline src/front/wgsl/lower/mod.rs:2452]: You probably don't want to use `InvalidAtomicPointer` here.

File: src/back/spv/writer.rs [hunk 4]
```diff
@@ -961,9 +973,9 @@ impl Writer {
 
                     self.write_type_declaration_local(id, local);
 
-                    // If it's an image type, request SPIR-V capabilities here, so
-                    // write_type_declaration_local can stay infallible.
-                    self.request_image_capabilities(&ty.inner)?;
+                    // If it's an type that needs SPIR-V capabilities, request them now,
```
Comment (jimblandy) [inline src/back/spv/writer.rs:976]:
```suggestion
                    // If it's a type that needs SPIR-V capabilities, request them now,
```

**Review by jimblandy — CHANGES_REQUESTED** [review]
This is interesting. I don't fully understand the ray tracing API, but I've tried to follow what's going on as best I could. Just a few points to address.

File: src/lib.rs [hunk 5]
```diff
@@ -1743,6 +1778,19 @@ pub struct EntryPoint {
...
+#[derive(Debug, Default)]
+#[cfg_attr(feature = "clone", derive(Clone))]
+#[cfg_attr(feature = "serialize", derive(Serialize))]
+#[cfg_attr(feature = "deserialize", derive(Deserialize))]
+#[cfg_attr(feature = "arbitrary", derive(Arbitrary))]
+pub struct SpecialTypes {
+    /// Type for `RayDesc`.
+    ray_desc: Option<Handle<Type>>,
+    /// Type for `RayIntersection`.
+    ray_intersection: Option<Handle<Type>>,
```
Comment (jimblandy) [inline src/lib.rs:1791]: Should these fields be `pub` too? Generally we want to support third-party producers and consumers of `Module`.

**Review by jimblandy — CHANGES_REQUESTED** [review]

**Comment by jimblandy** [comment]
@kvark By the way, I went ahead and pushed a commit that expands on the documentation, so be sure to pull before working on the branch. If the docs aren't accurate, feel free to just take the commit out.

File: src/valid/function.rs [hunk 6]
```diff
@@ -801,6 +801,9 @@ impl super::Validator {
                 } => {
                     self.validate_atomic(pointer, fun, value, result, context)?;
                 }
+                S::RayQuery { query: _, fun: _ } => {
+                    //TODO
```
[the remark this answers is not in the document]
Comment (kvark) [reply src/valid/function.rs:805]: I had actually implemented validation in #2282 follow-up, now cherry-picking it here

File: src/front/wgsl/lower/mod.rs [hunk 7]
```diff
@@ -644,14 +637,21 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
...
+        for decl_handle in self.index.visit_ordered() {
+            let span = tu.decls.get_span(decl_handle);
+            let decl = &tu.decls[decl_handle];
+
+            //TODO: find a nicer way?
+            if let Some(dep) = decl.dependencies.iter().find(|dep| dep.ident == "RayDesc") {
+                let ty_handle = ctx.module.generate_ray_desc_type();
+                ctx.globals
+                    .insert(dep.ident, LoweredGlobalDecl::Type(ty_handle));
+            }
```
[the remark this answers is not in the document]
Comment (kvark) [reply src/front/wgsl/lower/mod.rs:649]: Adding a comment. The issue can be seen on the `ray-query.wgsl` test if you comment this block out.

File: src/back/mod.rs [hunk 8]
```diff
@@ -218,3 +218,26 @@ impl crate::Statement {
...
+bitflags::bitflags! {
+    /// Ray flags.
+    #[derive(Default)]
+    pub struct RayFlag: u32 {
+        const OPAQUE = 0x01;
+        const NO_OPAQUE = 0x02;
+        const TERMINATE_ON_FIRST_HIT = 0x04;
+        const SKIP_CLOSEST_HIT_SHADER = 0x08;
+        const CULL_FRONT_FACING = 0x10;
+        const CULL_BACK_FACING = 0x20;
```
[the remark this answers is not in the document]
Comment (kvark) [reply src/back/mod.rs:238]: Nicely spotted! Unsure how I got this wrong 😅

File: src/valid/interface.rs (hunk 1 above)
Comment (kvark) [reply src/valid/interface.rs:445]: I didn't test this comprehensively. I assume they can be private globals too.

**Comment by kvark** [comment]
@jimblandy thanks for the review! Everything should be addressed now.
I'd also like to hear your thoughts on naming, i.e. `ray_desc` vs `RayDesc`.

**Review by jimblandy — APPROVED** [review]
Looks good to me. Give my `RayDesc` fix a look, and then feel free to merge if it seems okay.

**Comment by kvark** [comment]
Much nicer than my version, thank you!

**Pull request merged**